cip/CIP2016-06-14-Redefine-comparability-and-equality-as-well-as-orderability-and-equivalence.adoc

```diff
@@ -0,0 +1,367 @@
+* `avg(expr)` returns the arithmetic mean of aggregated values, or `0` if the list of aggregated values is empty.
+* `collect(expr)` returns the list of aggregated values.
+* `stdev(expr)` returns the standard deviation of the aggregated values (assuming they represent a random sample), or `0` if the list of aggregated values is empty.
+* `stdevp(expr)` returns the standard deviation of the aggregated values (assuming they form a complete population), or `0` if the list of aggregated values is empty.
+* `percentile_disc(expr)` computes the inverse distribution function (assuming a discrete distribution model), or `0` if the list of aggregated values is empty.
+* `percentile_cont(expr)` computes the inverse distribution function (assuming a continous distribution model), or `0` if the list of aggregated values is empty.
+
+[[conceptual-model]]
+=== Summary of the conceptual model ===
+
+This proposal aims to simplify the conceptual model around equality, comparison, order, and grouping:
+
+- <<comparability-equality,Comparability and equality>> are aligned with each other
+  * <<equality-con,Equality>> follows natural, literal equality. However, values involving `null` are never equal to any other value. Nested structures are first tested for equality by shape (keys, size) and then their corresponding elements are tested for equality pairwise. This ensures that equality is compatible with interpreting `null` as "unknown" or "could be any value".
+  * <<comparability-con,Comparability>> ensure that any two values of the same leaf type in the <<global-sort-order>> are comparable.
+  Two values of different types are incomparable and values involving `null` are incomparable, too.
+  This ensures that `MATCH (n) WHERE n.prop < 42` will never find nodes where `n.prop` is of type `STRING`.
+- <<orderability-equivalence>> are aligned with each other
+  * <<equivalence-con,Equivalence>> is a form of equality that treats `null` (and `NaN`) values as the same value. Equivalence is used in grouping and `DISTINCT` where `null` commonly is interpreted as a category marker for results with missing values instead of as a wildcard for any possible value.
+  * <<orderability-con,Orderability>> follows comparability but additionally defines a <<global-sort-order>> between values of different types and is aligned with equivalence instead of equality, i.e. treats two `null` (resp. `NaN`) values as equivalent.
+- Aggregation functions that rely on notions of sameness and sorting are aligned with equivalence and orderability.
+
+[[examples]]
+=== Examples
+
+An integer compared to a float
+[source, cypher]
+----
+RETURN 1 > 0.5 // should be true
+----
+
+A string compared to a boolean
+[source, cypher]
+----
+RETURN 'string' <= true // should be null
+----
+
+Ordering values of different types
+[source, cypher]
+----
+UNWIND [1, true, '', 3.14, {}, [2]] AS i
+RETURN i
+  ORDER BY i // should not crash
```

Comment:
We can do better than "should not crash", given the orderability definition, we should be able to tell the exact order in which these items are returned.